[RISCV]: Add support for RISCV instructions

lukeg101 opened the pull request:
This patch adds support for RISCV instructions I have encountered whilst testing compilers. Motivations are in the commits for each one.

Tests are included which all pass in herd, but I do not have RISCV hardware so I cannot test litmus - I am trying to figure out how to run tests on QEMU.

lukeg101 commented:
Apologies - rebase issues (please ignore)

lukeg101 commented:
I have found an 'easy' way to test RISCV programs using litmus on MacOS (documenting here for future people :) ) 
requires docker but somebody has embedded QEMU `qemu-static` into a Dockerfile
1. `mkdir out`
2. `litmus7 -o out @all` (where `@all` is your list of RISCV litmus tests)
3.  ` docker run --platform linux/riscv64 -v $(pwd):/root/test -it drujensen/riscv-ubuntu bash` (this person actively maintains this)
4. (in docker)# `cd test/out`
5. `sh ./comp.sh`
6. `sh ./run.sh`

lukeg101 commented:
Of interest to @maranget , I am looking into whether this is a bug in `risk.cat` now.
```
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/RISCV/A009.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RISCV A009

{
 0:x6=x; 0:x8=y;
 1:x6=y; 1:x8=x;
}
 P0          | P1          ;
 ori x5,x0,1 | ori x5,x0,1 ;
 sw x5,0(x6) | sw x5,0(x6) ;
 fence rw,w  | fence rw,w  ;
 lw x7,0(x8) | lw x7,0(x8) ;

exists (0:x7=0 /\ 1:x7=0)
Generated assembler
#START _litmus_P1
        ori t1,x0,1
        sw t1,0(a5)
        fence rw,w
        lw a1,0(a4)
#START _litmus_P0
        ori t1,x0,1
        sw t1,0(a5)
        fence rw,w
        lw a1,0(a4)
Test A009 Allowed
Histogram (3 states)
499993:>0:x7=1; 1:x7=0;
500000:>0:x7=0; 1:x7=1;
7     :>0:x7=1; 1:x7=1;
No

Witnesses
Positive: 0, Negative: 1000000
Condition exists (0:x7=0 /\ 1:x7=0) is NOT validated
Hash=247e5db6d71f2443e20d8676915c4192
Observation A009 Never 0 1000000
```

maranget commented:
If I understand correctly, I do not think so:
  1. The model allows the behaviour, the behaviour is not observed, so far so good.
  2. The litmus run is on top of qemu ? If so there is no memory model simulation, and one cannot deduce anything regarding the memory model from these runs.

maranget commented:
BTW, installing qemu and a riscv toolchain is rather easy on a linux box:
```
qemu     : sudo apt install qemu-user
toolchain: sudo apt install gcc-riscv64-linux-gnu
```
Compile with this litmus configuration file, [riscv-qemu.cfg](https://github.com/herd/herdtools7/blob/master/litmus/libdir/riscv-qemu.cfg). You may have to change the `gcc` setting to `riscv64-linux-gnu-gcc`.

For compiling proceed as if compiling and running on the same machine:
```
% mkdir R
% litmus7 -mach riscv-qemu -o R [options] [litmus files]
% cd R
% make
% sh ./run.sh [options]
```
Have a look at `run.sh` for the exact call of qemu.

lukeg101 commented:
Thanks for providing that Luc :) I am away until Sunday night but will come back and check using your instructions. I have similar testing for PPC and MIPS which I presume is similar (I have cross compilers for all archs I am testing)

lukeg101 commented:
Hi Luc,
Thank you for your help, attached are the litmus results:
```
Mon Aug 21 12:55:03 PM BST 2023
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/RISCV/A001.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RISCV A001

{
 [x]=0;
}
 P0  ;
 nop ;

forall ([x]=0)
Generated assembler
#START _litmus_P0
	nop
Test A001 Required
Histogram (1 states)
2000000:>[x]=0;
Ok

Witnesses
Positive: 2000000, Negative: 0
Condition forall ([x]=0) is validated
Hash=c4a3be38342b6b819f0d792f2ec53119
Observation A001 Always 2000000 0
Time A001 0.42
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/RISCV/A003.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RISCV A003

{
 0:x10=4;
}
 P0             ;
 srli x10,x10,1 ;

forall (0:x10=2)
Generated assembler
#START _litmus_P0
	srli a4,a4,1
Test A003 Required
Histogram (1 states)
2000000:>0:x10=2;
Ok

Witnesses
Positive: 2000000, Negative: 0
Condition forall (0:x10=2) is validated
Hash=0f76c62e95e0af150e8776b7b0859e14
Observation A003 Always 2000000 0
Time A003 0.27
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/RISCV/A004.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RISCV A004

{
 0:x11=1;
}
 P0         ;
 mv x10,x11 ;

forall (0:x10=1)
Generated assembler
#START _litmus_P0
	mv a4,a6
Test A004 Required
Histogram (1 states)
2000000:>0:x10=1;
Ok

Witnesses
Positive: 2000000, Negative: 0
Condition forall (0:x10=1) is validated
Hash=e78e73fbb56c1ec96e58fdadbb61f9c0
Observation A004 Always 2000000 0
Time A004 0.26
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/RISCV/A005.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RISCV A005

{
}
 P0  ;
 nop ;

forall (0:x10=0)
Generated assembler
#START _litmus_P0
	nop
Test A005 Required
Histogram (1 states)
2000000:>0:x10=0;
Ok

Witnesses
Positive: 2000000, Negative: 0
Condition forall (0:x10=0) is validated
Hash=7dbd8e4fdafc2b299a8e4b02a379dcaa
Observation A005 Always 2000000 0
Time A005 0.26
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/RISCV/A006.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RISCV A006

{
 0:x11=2;
}
 P0             ;
 sext.w x10,x11 ;

forall (0:x10=2)
Generated assembler
#START _litmus_P0
	sext.w a4,a6
Test A006 Required
Histogram (1 states)
2000000:>0:x10=2;
Ok

Witnesses
Positive: 2000000, Negative: 0
Condition forall (0:x10=2) is validated
Hash=b34f92984fa6bea21a79bed7c0a0765e
Observation A006 Always 2000000 0
Time A006 0.26
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/RISCV/A007.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RISCV A007

{
 0:x10=0;
}
 P0          ;
 auipc x10,2 ;

forall (0:x10=1)
Generated assembler
#START _litmus_P0
	auipc a4,2
Test A007 Required
Histogram (1 states)
2000000*>0:x10=90052;
No

Witnesses
Positive: 0, Negative: 2000000
Condition forall (0:x10=1) is NOT validated
Hash=4261d2fe6f90dd9cb08f7c74a9aead21
Observation A007 Never 0 2000000
Time A007 0.27
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/RISCV/A008.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RISCV A008

{
 0:x10=0;
}
 P0           ;
 lui x10,x0,1 ;

forall (0:x10=4096)
Generated assembler
#START _litmus_P0
	lui a4,1
Test A008 Required
Histogram (1 states)
2000000:>0:x10=4096;
Ok

Witnesses
Positive: 2000000, Negative: 0
Condition forall (0:x10=4096) is validated
Hash=49959c5922b8ac125416366e45fa5b62
Observation A008 Always 2000000 0
Time A008 0.27
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/RISCV/A009.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RISCV A009

{
 0:x6=x; 0:x8=y;
 1:x6=y; 1:x8=x;
}
 P0          | P1          ;
 ori x5,x0,1 | ori x5,x0,1 ;
 sw x5,0(x6) | sw x5,0(x6) ;
 fence rw,w  | fence rw,w  ;
 lw x7,0(x8) | lw x7,0(x8) ;

exists (0:x7=0 /\ 1:x7=0)
Generated assembler
#START _litmus_P1
	ori t3,x0,1
	sw t3,0(a5)
	fence rw,w
	lw a1,0(a4)
#START _litmus_P0
	ori t3,x0,1
	sw t3,0(a5)
	fence rw,w
	lw a1,0(a4)
Test A009 Allowed
Histogram (3 states)
499994:>0:x7=1; 1:x7=0;
499990:>0:x7=0; 1:x7=1;
16    :>0:x7=1; 1:x7=1;
No

Witnesses
Positive: 0, Negative: 1000000
Condition exists (0:x7=0 /\ 1:x7=0) is NOT validated
Hash=247e5db6d71f2443e20d8676915c4192
Observation A009 Never 0 1000000
Time A009 0.81
Revision 99ddbe4ec5d833fdf4a3e06a128427f9159363f1, version 7.56+03
Command line: ./_build/install/default/bin/litmus7 -mach litmus/libdir/riscv-qemu.cfg -set-libdir litmus/libdir -o out herd/tests/instructions/RISCV/@all
Parameters
#define SIZE_OF_TEST 5000
#define NUMBER_OF_RUN 200
#define AVAIL 2
#define STRIDE 1
#define MAX_LOOP 0
/* gcc options: -Wall -std=gnu99 -O2 -pthread */
/* gcc link options: -static */
/* barrier: user */
/* launch: changing */
/* affinity: none */
/* memory: direct */
/* stride: 1 */
/* safer: write */
/* preload: random */
/* speedcheck: no */
/* alloc: dynamic */
/* proc used: 2 */
GCC=riscv64-unknown-linux-gnu-gcc
LITMUSOPTS=
Mon Aug 21 12:55:06 PM BST 2023
```

Some notes: 
- `A007` is incorrect as it requires a PC relative calculation which is not supported in litmus (like ADRP)
- `A009` is allowed by the model but not observed here

I am inclined to mark this as ready to review now

lukeg101 commented:
Open question: Shall I mark `AUIPC` as not supported in litmus (I still need it for parsing compiled programs)

maranget commented:
> Open question: Shall I mark `AUIPC` as not supported in litmus (I still need it for parsing compiled programs)

Is `AUIPC` supported in herd?

lukeg101 commented:
I did not support it in herd in the above patch as it is like ADRP in AArch64 - I recall we decided not to support ADRP as we could not test it in litmus. I am doing the same here with RISCV and AUIPC (it is used to compute addresses in LLVM/GCC code). Is this ok?

maranget commented on herd/RISCVSem.ml:
```diff
@@ -255,7 +255,8 @@ module
           begin match ii.A.inst with
-          | RISCV.INop -> B.next1T ()
+          | RISCV.INop-> B.next1T ()
+          | RISCV.Ret -> M.unitT () >>! B.Exit
```
As far as I could check, RISCV RET is jump tp the content of `ra` (similar to AArch64 ret).

lukeg101 replied:
Guarding under the `Telechat` variant

maranget commented on herd/RISCVSem.ml:
```diff
@@ -255,10 +257,20 @@ module
           begin match ii.A.inst with
+          | RISCV.INop-> B.next1T ()
+          | RISCV.Ret -> M.unitT () >>! B.Exit
+          | RISCV.OpI (RISCV.LUI,r1,_,k) ->
```
This encoding of the `LUI` instruction is surprising. Moreover the instruction `lui r,k` is printed as `lui r,x0,k`.

maranget commented on lib/RISCVBase.ml:
```diff
@@ -286,18 +291,23 @@ type signed = Sign.t
   | OpW of opw * reg * reg * reg
   | J of lbl
   | Bcc of cond * reg * reg * lbl
+  | MV of reg * reg
```
`mv` is a pseudoinstruction (`mv rd,rs` being `addi rd,rs,0`). It would be natural not to have  specific abstract instruction. Instead one could parse `mv rd,rs` as the its expansion. Then one should probably print `addi rd,rs,0` as `mv rd,rs` and also emit the `mv` form as litmus output. However those last two steps are optional, although recommended.

I know that some other pseudoinstructions are not implemented that way (AArch64 `li r,k` for instance). However, it would be quite easy to do so in the case of RISCV `mv`, given that `addi` is already available.

lukeg101 replied:
removed mv constructor, pretty print `add rd, rs, 0` as `mv rd,rs`, parse mv as add.

maranget commented:
> I did not support it in herd in the above patch as it is like ADRP in AArch64 - I recall we decided not to support ADRP as we could not test it in litmus. I am doing the same here with RISCV and AUIPC (it is used to compute addresses in LLVM/GCC code). Is this ok?

I guess this is ok. If AUIPC does not mean much in litmus executable, perhaps it is better not emit this instruction and to flag a fatal error.

maranget commented:
As regards the new arguments of the fence instruction, I could find no reference. I found `wo` in place of `ow` and I guess `ir` may also exists.... But I am not sure of anything here.

maranget commented on lib/RISCVParser.mly:
```diff
@@ -17,10 +17,11 @@
   | "r" -> A.R
-  | "w" -> A.W
-  | "rw" -> A.RW
+  | "w" | "ow" -> A.W
+  | "rw" | "iorw" -> A.RW
```
As a result of this coding `iorw` will be printed as `rw` and litmus will emit `rw` in place of `iorw`... Is this intended?

lukeg101 replied:
On table 26.3 of the [reference](https://riscv.org/wp-content/uploads/2019/06/riscv-spec.pdf) I can see that `fence iorw,iorw` means `Fence on all memory and I/O`. I could not find a reference for the meaning of `ow`. Since we only deal with Memory in herd I treat `iorw` as `rw`. Is this ok?

lukeg101 commented:
> I guess this is ok. If AUIPC does not mean much in litmus executable, perhaps it is better not emit this instruction and to flag a fatal error.
Marked as 'not supported' in litmus

lukeg101 commented:
Hi Luc, thank you very much for reviewing. I have addressed your comments above, rebased onto master, and added the test you kindly provided. Please let me know if there is anything else.
Thanks
Luke

maranget commented:
Hi @lukeg101. Ready for merge, if you do not object.

maranget commented:
Merged, thanks @lukeg101.
